### ISSUE TITLE
Serialize/Deserialize a struct as a tuple

### ISSUE THREAD

**Issue by beatgammit** [serde-rs/serde#297]
I'm toying with implementing a protocol that sends tuples across the wire. Each of the elements of the tuple has a distinct meaning, and it's useful to represent it as a struct in client code. I haven't looked at the bincode implementation, but it seems to do something similar.

For example, if I have a struct like this:

    struct A {
        x: i32,
        y: f64,
    }

It would be serialized as `(i32, f64)` with each index. This saves on space by omitting the field names.

The format that I'm interested in representing also tries to eliminate extra fields when unnecessary, so if something is None, it would be left off if it's at the end of the struct. For example:

    struct A {
        x: i32,
        y: (),
    }

Would be serialized as `(i32)`.

Now, this second part makes serde more complicated as it needs to know not only if the next element is None, but if all of the following elements are none all the way to the end of the struct. This feature isn't useful to me if I only get the first part, so I'm interested in knowing if this idea is feasible given serde's goals, or if this should just be done in application code instead.

I'm somewhat new to rust, but in Go I used reflection to generate an array. If you decide that this feature is out of scope, could you suggest an idiomatic approach to use in Rust? I fear for my particular case I'll have to manually serialize my structures =(

**Comment by dtolnay** [comment]
My impression is that this is probably not a common enough usage to be valuable supporting first-class in Serde. It is simple enough to implement the serialization yourself with this behavior. The code could look something like this:

```rust
impl Serialize for A {
    fn serialize<S: Serializer>(&self, s: &mut S) -> Result<(), S::Error> {
        if self.z.is_some() {
            (&self.x, &self.y, &self.z).serialize(s)
        } else if self.y.is_some() {
            (&self.x, &self.y).serialize(s)
        } else if self.x.is_some() {
            (&self.x,).serialize(s)
        } else {
            ().serialize(s)
        }
    }
}
```

You could define your own macro to generate the above, as in:

```rust
beatgammit_serialize!(A { x, y, z })
```

**Comment by dtolnay** [comment]
Here is one possible way to implement the `beatgammit_serialize!` macro to generate implementations like the one in my previous comment:

```rust
macro_rules! beatgammit_serialize {
    // helper: recursive case
    (($s:ident), ($($prev:expr),*), ($cur:expr $(, $rest:expr)*), $tail:expr) => {
        beatgammit_serialize!(
            ($s),
            ($($prev ,)* $cur),
            ($($rest),*),
            if $cur.is_some() {
                ($(&$prev ,)* &$cur,).serialize($s)
            } else {
                $tail
            })
    };
    // helper: base case
    (($s:ident), ($($prev:expr),*), (), $tail:expr) => {
        $tail
    };
    // beatgammit_serialize!(A { x, y, z })
    ($ty:ident { $($field:ident),* }) => {
        impl Serialize for $ty {
            fn serialize<S: Serializer>(&self, s: &mut S) -> Result<(), S::Error> {
                beatgammit_serialize!((s), (), ($(self.$field),*), ().serialize(s))
            }
        }
    };
}

beatgammit_serialize!(A { x, y, z });
```

**Comment by beatgammit** [comment]
Thanks! I was thinking it would have to be something like this, but I kept thinking I had to return the tuple and let the caller serialize it, but I'm glad to see there was a trait for this (didn't see it right off).

And yes, this is a relatively fringe case, however I'd guess a binary serialization like capnproto or bincode would benefit as well. It's certainly not worth it if implementing negatively affects everything else.

Thanks for your help!
